controllers/release/release_adapter.go

```diff
@@ -0,0 +1,306 @@
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package release
+
+import (
+	"context"
+	"fmt"
+	"github.com/go-logr/logr"
+	libhandler "github.com/operator-framework/operator-lib/handler"
+	"github.com/redhat-appstudio/release-service/api/v1alpha1"
+	"github.com/redhat-appstudio/release-service/controllers/results"
+	"github.com/redhat-appstudio/release-service/tekton"
+	"github.com/tektoncd/pipeline/pkg/apis/pipeline/v1beta1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/types"
+	"knative.dev/pkg/apis"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller/controllerutil"
+	"time"
+)
+
+// Adapter holds the objects needed to reconcile a Release.
+type Adapter struct {
+	release *v1alpha1.Release
+	logger  logr.Logger
+	client  client.Client
+	context context.Context
+}
+
+// releaseFinalizer is the finalizer name to be added to the Releases
+const releaseFinalizer string = "appstudio.redhat.com/finalizer"
+
+// NewAdapter creates and returns an Adapter instance.
+func NewAdapter(release *v1alpha1.Release, logger logr.Logger, client client.Client, context context.Context) *Adapter {
+	return &Adapter{
+		release: release,
+		logger:  logger,
+		client:  client,
+		context: context,
+	}
+}
+
+// EnsureFinalizersAreCalled is an operation that will ensure that finalizers are called whenever the Release being
+// processed is marked for deletion.
+func (a *Adapter) EnsureFinalizersAreCalled() (results.OperationResult, error) {
+	if a.release.GetDeletionTimestamp() != nil {
+		if controllerutil.ContainsFinalizer(a.release, releaseFinalizer) {
+			if err := a.finalizeRelease(); err != nil {
+				return results.RequeueWithError(err)
+			}
+
+			controllerutil.RemoveFinalizer(a.release, releaseFinalizer)
+			err := a.client.Update(a.context, a.release)
+			if err != nil {
+				return results.RequeueWithError(err)
+			}
+		}
+
+		return results.Requeue()
+	}
+
+	return results.ContinueProcessing()
+}
+
+// EnsureFinalizerIsAdded is an operation that will ensure that the Release being processed contains a finalizer.
+func (a *Adapter) EnsureFinalizerIsAdded() (results.OperationResult, error) {
+	var finalizerFound bool
+	for _, finalizer := range a.release.GetFinalizers() {
+		if finalizer == releaseFinalizer {
+			finalizerFound = true
+		}
+	}
+
+	if !finalizerFound {
+		a.logger.Info("Adding Finalizer to the Release")
+		controllerutil.AddFinalizer(a.release, releaseFinalizer)
+		err := a.client.Update(a.context, a.release)
+
+		return results.RequeueOnErrorOrContinue(err)
+	}
+
+	return results.ContinueProcessing()
+}
+
+// EnsureReleasePipelineRunExists is an operation that will ensure that a release PipelineRun associated to the Release
+// being processed exists.
```

Comment:
I think you should mention what the function does if the pipeline run does not exist when this is called